File: 4.model_reader:shader_template:sphere/57.SphereExample3/app/src/main/java/com/example/martinrgb/a57sphereexample3/OpenGLView.java
package com.example.martinrgb.a57sphereexample3;

import android.content.Context;
import android.opengl.GLSurfaceView;
import android.view.MotionEvent;

public class OpenGLView extends GLSurfaceView {

    private OpenGLRenderer4 mRenderer;

    private float mDownX = 0.0f;
    private float mDownY = 0.0f;

    public OpenGLView(Context context) {
        super(context);

        mRenderer = new OpenGLRenderer4();
        this.setRenderer(mRenderer);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        int action = event.getActionMasked();
        switch (action) {
            case MotionEvent.ACTION_DOWN:
                mDownX = event.getX();
                mDownY = event.getY();
                return true;
            case MotionEvent.ACTION_UP:
                return true;
            case MotionEvent.ACTION_MOVE:
                float mX = event.getX();
                float mY = event.getY();
                mRenderer.mLightX += (mX-mDownX)/10;
                mRenderer.mLightY -= (mY-mDownY)/10;
                mDownX = mX;
                mDownY = mY;
                return true;
            default:
                return super.onTouchEvent(event);
        }
    }
}

File: 4.model_reader:shader_template:sphere/57.SphereExample3/57.SphereExample3_kotlin/app/src/main/java/com/example/martinrgb/a57sphereexample3/OpenGLView.kt
package com.example.martinrgb.a57sphereexample3

import android.content.Context
import android.opengl.GLSurfaceView
import android.view.MotionEvent

class OpenGLView(context: Context) : GLSurfaceView(context) {

    private val mRenderer: OpenGLRenderer4

    private var mDownX = 0.0f
    private var mDownY = 0.0f

    init {

        mRenderer = OpenGLRenderer4()
        this.setRenderer(mRenderer)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        val action = event.actionMasked
        when (action) {
            MotionEvent.ACTION_DOWN -> {
                mDownX = event.x
                mDownY = event.y
                return true
            }
            MotionEvent.ACTION_UP -> return true
            MotionEvent.ACTION_MOVE -> {
                val mX = event.x
                val mY = event.y
                mRenderer.mLightX = mRenderer.mLightX + (mX - mDownX) / 10
                mRenderer.mLightY = mRenderer.mLightY - (mY - mDownY) / 10
                mDownX = mX
                mDownY = mY
                return true
            }
            else -> return super.onTouchEvent(event)
        }
    }
}